multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,299 @@
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if req.NamespacedName.Namespace == "" {
+		err := r.Client.Get(ctx, req.NamespacedName, &ns)
+		if err == nil {
+			if ns.DeletionTimestamp.IsZero() {
+				// Based on the predicates used to register the reconciler, this can only be a
+				// Namespace add or Namespace label update event
+				return ctrl.Result{}, r.onNamespaceCreateOrUpdate(ctx, req.Name, ns.Labels)
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			return ctrl.Result{}, r.onPodDelete(ctx, req.NamespacedName.String())
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	nsLabels, podLabels := ns.Labels, pod.Labels
+	return ctrl.Result{}, r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), getNormalizedLabel(nsLabels, podLabels))
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Ignore status update event via GenerationChangedPredicate
+	instance := predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(instance).
+		Watches(&source.Kind{Type: &v1.Namespace{}},
+			handler.EnqueueRequestsFromMapFunc(namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc enqueues Namespace events to the reconciler.
+func namespaceMapFunc(o client.Object) []reconcile.Request {
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Name:      o.GetName(),
+				Namespace: o.GetNamespace(),
+			},
+		},
+	}
+}
+
+// onNamespaceCreateOrUpdate updates the label identities cached for all Pods in
+// the Namespace in case the Namespace's labels change
+func (r *LabelIdentityReconciler) onNamespaceCreateOrUpdate(ctx context.Context, namespace string, nsLabels map[string]string) error {
+	var pods v1.PodList
+	if err := r.Client.List(ctx, &pods, &client.ListOptions{Namespace: namespace}); err != nil {
+		klog.Errorf("Failed to list Pods in Namespace %s", namespace)
+		return err
+	}
+	for _, p := range pods.Items {
+		podKey, updatedLabel := p.Namespace+"/"+p.Name, getNormalizedLabel(nsLabels, p.Labels)
+
+		klog.V(4).InfoS("Re-processing Pod for label identity sync due to Namespace label updates", "pod", podKey)
+		if err := r.onPodCreateOrUpdate(ctx, podKey, updatedLabel); err != nil {
```

Comment:
Is this safe without synchronization with Pod events? For example, how if we get stale Pod labels here, but due to race condition onPodCreateOrUpdate() is called later than that of a Pod update event?

Reply:
Agreed that it might not be safe. Refactored to have mutex protection between any Namespace and Pod events.

Reply:
I was thinking should we generate and queue Pod events here, and so we have a single thread to handle events for a single Pod. @tnqn : should give better suggestion here.

Reply:
@jianjuns Changed this controller to use a pod event queue. Please check if this fits what you were thinking.

Reply:
I'm not familiar with controller-runtime, but I feel the way label identify controller uses it seems not matching its interface's design:
The facts that the argument of `Reconcile` method doesn't have any type information and `SetupWithManager` calls `For(&v1.Pod{})` to declare Pod is the Object being *reconciled* mean that, the function is supposed to be used to reconcile Pod only, not mixed for Pod and Namespace. It seems the code reuses the queue for both Namespaces and Pods with a trick and differentiates types by checking whether Namespace is empty. 
From the implementation of control-runtime/pkg/internal/controller/controller.go, the `Reconcile` method is already what "worker" executes (the controller library already sets up watcher, eventhandler, workqueue, workers), so it doesn't make a lot of sense to enqueue the key again and spawn another group of workers to actually handle it, which makes using the control-runtime library pointless.

I understand the requirement is, we reconcile Pod object only, but both Pod and Namespace are the source of truth, so we need to watch Namespace and make sure when a Namespace's label is updated, corresponding Pods should be reconciled. I guess `Watches(src source.Source, eventhandler handler.EventHandler, opts ...WatchesOption)` is designed by this requirement, but instead of enqueuing the Namespace object itself to the queue of Pod, I think the registered eventhandler should list all Pods in that Namespace and return them as "requests", something like:
```
Watches(&source.Kind{Type: &v1.Namespace{}}, handler.EnqueueRequestsFromMapFunc(func(ns client.Object) []reconcile.Request {
	podList := &v1.PodList{}
	mgr.GetClient().List(context.TODO(), podList, client.InNamespace(ns.GetName()))
	requests := make([]reconcile.Request, len(podList.Items))
	for _, pod := range podList.Items {
		requests = append(requests, reconcile.Request{
			NamespacedName: types.NamespacedName{
				Name:      pod.GetName(),
				Namespace: pod.GetNamespace(),
			},
		})
	}
	return requests
})).
```

Reply:
I am not familiar with controller-runtime. But what Quan described sounds quite reasonable to explain the interfaces.
@luolanzone may have comments too.

Reply:
Updated to use the requeue logic that Quan suggested